File: lab1_Java/src/Calculator.java
public class Calculator
{
    private long _sum;
    private int _threadsCount;
    private int _endedThreads;

    public long sumInMultipleThreads(int[] array, int threadsCount)
    {
        _sum = 0;
        _threadsCount = threadsCount;
        _endedThreads = 0;

        PartialCalculator[] partialCalculators = new PartialCalculator[threadsCount];

        for (int i = 0; i < threadsCount; i++)
        {
            partialCalculators[i] = new PartialCalculator(this, array, i, threadsCount);
            partialCalculators[i].start();
        }

        waitForEndSumCalculation();

        return _sum;
    }

    synchronized public void addSum(long value)
    {
        _sum += value;
        _endedThreads++;
        notify();
    }

    synchronized public void waitForEndSumCalculation()
    {
        while(_endedThreads < _threadsCount)
        {
            try
            {
                wait();
            }
            catch (InterruptedException e)
            {
                e.printStackTrace();
            }
        }
    }
}

File: lab1_Csharp/Calculator.cs
using System.Diagnostics;
using System.Threading;

namespace lab1_Csharp
{
    public class Calculator
    {
        public long SumInMultipleThreads(int[] array, int threadsCount, out float calculationTime)
        {
            long sum = 0;

            Thread[] threads = new Thread[threadsCount];
            PartialCalculator[] partialCalculators = new PartialCalculator[threadsCount];

            Stopwatch stopwatch = new Stopwatch();
            stopwatch.Start();

            for (int i = 0; i < threads.Length; i++)
            {
                partialCalculators[i] = new PartialCalculator(array, i, threadsCount);
                threads[i] = new Thread(new ThreadStart(partialCalculators[i].Calculate));
                threads[i].Start();
            }

            for (int i = 0; i < threads.Length; i++)
            {
                threads[i].Join();
                sum += partialCalculators[i].Sum;
            }

            stopwatch.Stop();
            calculationTime = (float)stopwatch.Elapsed.TotalMilliseconds;

            return sum;
        }
    }
}